bindings/java/src/main/com/apple/foundationdb/tuple/ByteArrayUtil.java

```diff
@@ -171,7 +173,69 @@ static boolean regionEquals(byte[] src, int start, byte[] pattern) {
 	 */
 	public static byte[] replace(byte[] src, int offset, int length,
 			byte[] pattern, byte[] replacement) {
-		return join(replacement, split(src, offset, length, pattern));
+		if(pattern == null || pattern.length == 0) {
+			return Arrays.copyOfRange(src, offset, offset + length);
+		}
+		ByteBuffer dest;
+		if(replacement == null || replacement.length != pattern.length) {
+			// Array might change size. This is the "tricky" case.
+			byte patternFirst = pattern[0];
+			int patternOccurrences = 0;
+			int currentPosition = offset;
+			while(currentPosition < offset + length) {
+				if(src[currentPosition] == patternFirst && regionEquals(src, currentPosition, pattern)) {
+					patternOccurrences++;
+					currentPosition += pattern.length;
+				}
+				else {
+					currentPosition++;
+				}
+			}
+			if(patternOccurrences == 0) {
+				// Pattern doesn't occur. Just return a copy of the needed region.
+				return Arrays.copyOfRange(src, offset, offset + length);
+			}
+			int replacementLength = (replacement == null) ? 0 : replacement.length;
+			int newLength = length + patternOccurrences * (replacementLength - pattern.length);
+			if(newLength == 0) {
+				return new byte[0];
+			}
+			else {
+				dest = ByteBuffer.allocate(newLength);
+			}
+		}
+		else {
+			// No matter what, the array will stay the same size as replacement.length = pattern.length
+			dest = ByteBuffer.allocate(length);
+		}
+		replace(src, offset, length, pattern, replacement, dest);
+		return dest.array();
+	}
+
+	static void replace(byte[] src, int offset, int length, byte[] pattern, byte[] replacement, ByteBuffer dest) {
+		if(pattern == null || pattern.length == 0) {
+			dest.put(src, offset, length);
+			return;
+		}
+		byte patternFirst = pattern[0];
+		int lastPosition = offset;
+		int currentPosition = offset;
+
+		while(currentPosition < offset + length) {
```

Comment:
Maybe it can write down all the start and end indices pairs for each occurrence to avoid the second search here.

Reply:
Yeah, I'm kind of unhappy *in general* with how these methods interact. It's possible the right thing to do is for the other (now unused) variant of replace to go back to the old code path (which is much simpler) rather than attempting to use this logic in the other one because it kind of leads to a lot of duplicated code, which is what I was trying to avoid (perhaps unsuccessfully).

Reply:
Okay, I tried to implement something like that. It's in its own commit for easy reverting if we don't like it.

Reply:
I'm happy with this